### ISSUE TITLE
In Internet Explorer Automation discuss BHOs

### ISSUE THREAD

**Issue by sancarn** [rubyonwindowsguides/rubyonwindowsguides.github.com#14]
Currently writing an article where I discuss the usage of BHOs on Windows in InternetExplorer which would be a great addition to this guide:

https://gist.github.com/sancarn/79042ba2036fd86bb23e00dc383cdcec#browser-helper-objects

The article also goes further into using chrome for GUIs, not sure if that'd also be useful but could be :)

**Comment by bosko** [comment]
BHOs are really interesting. I'm a bit away from Windows last couple years that's why I'm not updating this. What's your idea about adding this (you can mail me directly)?

**Comment by sancarn** [comment]
Basically I think it's mainly just important to highlight how you can use BHOs with IE to implement Ruby classes and objects into IE's Javascript. Take the following code:

```rb
# Create a new Internet Explorer COM object (use Late Binding PROGID/CLSID):
require 'win32ole'
ie = WIN32OLE.new('InternetExplorer.Application')

# Navigate to blank page
ie.navigate("about:blank")
sleep(0.1) until !ie.busy

html =<<-Heredoc
  <html>
    <head>
      <script>
        document.setVar = function(name,value){
          window[name] = value;
        }
      </script>
    </head>
  </html>
Heredoc

# Write HTML to page
ie.document.write(s)

#Make IE visible
ie.visible = true

# Extend the File object giving JS access to File API.
  class WIN32OLE_File < File
    # Allow calling of class methods
    def self.call
      nil
    end
    def self.value(*args)
      nil
    end

    # Allow calling of instance methods
    def call
      nil
    end
    def value(*args)
      nil
    end
  end
  
  #Inject WIN32OLE_File class into JavaScript engine
  ie.document.setVar("RubyFile",WIN32OLE_File)
```

After this code has been executed, you can now use Ruby's File object within JavaScript:

```js
var f = RubyFile.new("my/file","w")
f.write("hello world")
```

Of course this is important because more generally, the file system cannot be accessed directly with Internet Explorer and therefore allowing access via these Ruby scripts is extremely helpful!

**Comment by bosko** [comment]
I agree but it should be accompanied by some introductory text. Or maybe I can just add sentence or two with the link to your article when you publish it?

**Comment by sancarn** [comment]
If you like you can do so. It may be somewhat confusing that I talk about ICM all the time though. Otherwise I can port the current markdown to say "windows" instead of "ICM". Whichever makes more sense to you :)

P.S. The article itself is to be posted on LinkedIn so maybe not the best site to link to. Linking to the GIST is fine with me :) If thats the way you want to go with it.
